map_model/src/objects/block.rs

```diff
@@ -163,154 +159,156 @@ impl Perimeter {
         self.roads.push(self.roads[0]);
     }
 
-    /// Try to merge two blocks. Returns true if this is successful, which will only be when the
-    /// blocks are adjacent, but the merge wouldn't create an interior "hole".
-    ///
-    /// Note this may modify both perimeters and still return `false`. The modification is just to
-    /// rotate the order of the road loop; this doesn't logically change the perimeter.
+    /// Try to merge two blocks. This'll only succeed when the blocks are adjacent, but the merge
+    /// wouldn't create an interior "hole".
     ///
-    /// TODO Due to https://github.com/a-b-street/abstreet/issues/841, it seems like rotation
-    /// sometimes breaks `to_block`, so for now, always revert to the original upon failure.
-    // TODO Would it be cleaner to return a Result here and always restore the invariant?
+    /// Note this always modifies both perimeters, even upon failure. The caller should copy the
+    /// input and only use the output upon success.
     fn try_to_merge(
         &mut self,
         map: &Map,
         other: &mut Perimeter,
         debug_failures: bool,
-        use_expensive_blockfinding: bool,
-    ) -> bool {
-        let orig_self = self.clone();
-        let orig_other = other.clone();
-
-        self.undo_invariant();
-        other.undo_invariant();
+    ) -> Result<()> {
+        for reverse_to_fix_winding_order in [false, true] {
+            self.undo_invariant();
+            other.undo_invariant();
+
+            // Calculate common roads
+            let roads1: HashSet<RoadID> = self.roads.iter().map(|id| id.road).collect();
+            let roads2: HashSet<RoadID> = other.roads.iter().map(|id| id.road).collect();
+            let common: HashSet<RoadID> = roads1.intersection(&roads2).cloned().collect();
+            if common.is_empty() {
+                if debug_failures {
+                    warn!("No common roads");
+                }
+                bail!("No common roads");
+            }
+
+            // "Rotate" the order of roads, so that all of the overlapping roads are at the end of the
+            // list. If the entire perimeter is surrounded by the other, then no rotation needed.
+            if self.roads.len() != common.len() {
+                let mut i = 0;
+                while common.contains(&self.roads[0].road)
+                    || !common.contains(&self.roads.last().unwrap().road)
+                {
+                    self.roads.rotate_left(1);
+
+                    i += 1;
+                    if i == self.roads.len() {
```

Comment:
I don't know what's actually happening here. For now, just avoid infinite looping and forge ahead